lib/Terrain.cpp

```diff
@@ -164,12 +169,18 @@ Terrain::Manager & Terrain::Manager::get()
 	return manager;
 }
 
-std::vector<Terrain> Terrain::Manager::terrains()
+const std::vector<Terrain> & Terrain::Manager::terrains()
 {
-	std::set<Terrain> _terrains; //have to use std::set to have ordered container. Othervise de-sync is possible
-	for(const auto & info : Terrain::Manager::get().terrainInfo)
-		_terrains.insert(info.first);
-	return std::vector<Terrain>(_terrains.begin(), _terrains.end());
+	return Terrain::Manager::get().terrainVault;
+}
+
+int Terrain::Manager::id(const Terrain & terrain)
+{
+	if(terrain.name == "ANY") return -3;
+	if(terrain.name == "WRONG") return -2;
+	if(terrain.name == "BORDER") return -1;
+	
+	return Terrain::Manager::get().terrainId.at(terrain);
```

Comment:
can we put these special ids to terrainID map

Reply:
Certainly, but what this is a bit odd for me, but searching in map is longer than comparison (nevertheless I don't know frequency of usage)

What will be benefit? If it's readability only, I can make separate const array or something at the top of file.

Reply:
We will have one search only. The one at line 183.

Reply:
So this will be a bit odd I agree.